tests/test_references.py

```diff
@@ -186,13 +241,303 @@ async def test_deploy_end_to_end(
+    ### basic inheritance
+    run_snippet("refs::plugins::takes_no_refs_dataclass(refs::dc::MixedRefsDataclass(maybe_ref_value=refs::create_string_reference('hello')))")
+    ### basic inheritance the wrong direction
+    with pytest.raises(PluginTypeException, match=re.escape("Expected type: refs::dc::MixedRefsDataclass")):
+        run_snippet("refs::plugins::takes_mixed_refs_dataclass(refs::dc::NoRefsDataclass())")
+    ### references not allowed
+    #### references to dataclass
+    with pytest.raises(PluginTypeException, match="contains a reference"):
+        run_snippet("refs::plugins::takes_no_refs_dataclass(refs::dc::create_no_refs_dataclass_reference())")
+    with pytest.raises(PluginTypeException, match="contains a reference"):
+        run_snippet("refs::plugins::takes_mixed_refs_dataclass(refs::dc::create_mixed_refs_dataclass_reference('hello'))")
+    #### dataclass containing undeclared reference
+    with pytest.raises(PluginTypeException, match="contains a reference"):
+        run_snippet("refs::plugins::takes_no_refs_dataclass(refs::dc::NoRefsDataclass(non_ref_value=refs::create_string_reference('hello')))")
+    ## dataclass type that supports reference attrs -> references are coerced
+    ### references are coerced to dataclass of references
+    run_snippet("refs::plugins::takes_all_refs_dataclass(refs::dc::create_all_refs_dataclass_reference('hello'))")
+    ### references are coerced to dataclass of references, with inheritance
+    run_snippet("refs::plugins::takes_dataclass(refs::dc::create_all_refs_dataclass_reference('hello'))")
+    with pytest.raises(PluginTypeException, match="contains a reference"):
+        run_snippet("refs::plugins::takes_dataclass(refs::dc::create_no_refs_dataclass_reference())")
+
+    # Scenario: plugin annotated as Reference[<dataclass>]
+    ## accepts a reference
+    run_snippet("refs::plugins::takes_no_refs_dataclass_ref(refs::dc::create_no_refs_dataclass_reference())")
+    ## rejects a dataclass
+    with pytest.raises(PluginTypeException, match=re.escape("Expected type: Reference[refs::dc::NoRefsDataclass]")):
+        run_snippet("refs::plugins::takes_no_refs_dataclass_ref(refs::dc::NoRefsDataclass())")
+
+    # Scenario: plugin annotated as <dataclass> | Reference[<dataclass>]
+    ## accepts either
+    run_snippet("refs::plugins::takes_no_refs_dataclass_or_ref(refs::dc::create_no_refs_dataclass_reference())")
+    run_snippet("refs::plugins::takes_no_refs_dataclass_or_ref(refs::dc::NoRefsDataclass())")
+    run_snippet("refs::plugins::takes_mixed_refs_dataclass_or_ref(refs::dc::create_mixed_refs_dataclass_reference('hello'))")
+    run_snippet("refs::plugins::takes_mixed_refs_dataclass_or_ref(refs::dc::MixedRefsDataclass(maybe_ref_value='hello'))")
+    ## rejects other dataclasses / references to other dataclasses, except for inheritance
+    run_snippet("refs::plugins::takes_no_refs_dataclass_or_ref(refs::dc::MixedRefsDataclass(maybe_ref_value='hello'))")
```

Comment:
A would add a comment to indicate this is accepted because `MixedRefsDataclass` is a subclass of `NoRefsDataclass`

Reply:
Isn't that what the comment above this line says? Or do you mean to be more explicit?

Reply:
I'll expand on it